### PR TITLE
Use region specific API calls with VMSS

### LINKED ISSUE #1720: AKS test is flaky
/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

AKS tests have been failing very frequently recently with error "System machine pools not ready". Example:https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-provider-azure-e2e-full/1440295663572094976

dashboard: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-azure#capz-periodic-capi-e2e-full-main 

This is slowing down PRs to merge as they are unable to pass the full test suite.

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by devigned**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When listing VMSSes in a resource group, Azure responds with values it has in the regional cache. Unfortunately, there are sometimes spikes in regional replication for these caches. To avoid cross region replication issues, this PR introduces a region specific client for AzureManageMachinePools by transforming the client baseURI from something like `https://management.azure.com` to `https://{region}.management.azure.com` to ensure the request is directed to the region where the VMSSes should be.

fixes #1720

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
use a region specific Azure client for listing VMSS in AzureManagedClusters
```

### CODE REVIEW

**Comment by devigned** [comment]
success 2, 22mins

/test pull-cluster-api-provider-azure-e2e-exp

**Comment by devigned** [comment]
success 3, 23mins

/test pull-cluster-api-provider-azure-e2e-exp

File: exp/controllers/azuremanagedmachinepool_reconciler.go [hunk 1]
```diff
@@ -146,7 +153,11 @@ func (s *azureManagedMachinePoolService) Delete(ctx context.Context) error {
...
 
-// IsAgentPoolVMSSNotFoundError returns true if the error is an AgentPoolVMSSNotFoundError.
-func IsAgentPoolVMSSNotFoundError(err error) bool {
-	return errors.Is(err, notFoundErr)
+func getRegionalBaseURI(region, baseURI string) (string, error) {
+	parsedURI, err := url.Parse(baseURI)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to parse the base URI of client")
+	}
+
```
Comment (alexeldeib) [inline exp/controllers/azuremanagedmachinepool_reconciler.go:161]: Nice

**Comment by devigned** [comment]
success 4, 23mins

/test pull-cluster-api-provider-azure-e2e-exp

**Comment by alexeldeib** [comment]
23 min is right around what I’d expect for the total duration of this test. Looking good!

**Comment by alexeldeib** [comment]
Actually, with the latest bits to reconcile pools all together on cluster create, I wonder if we can’t squeeze that lower. Delete is kind of slow though, so we’ll see what makes sense.

**Comment by devigned** [comment]
5 passes, 5th 23mins

**Comment by devigned** [comment]
@CecileRobertMichon the PR history is now solid green for the exp job. I think we are solid. So solid that we may want to consider adding the AKS related tests to the regular pr e2e test.

**Comment by devigned** [comment]
> Actually, with the latest bits to reconcile pools all together on cluster create, I wonder if we can’t squeeze that lower. Delete is kind of slow though, so we’ll see what makes sense (unrelated to this PR)

I think with the transient error / requeue stuff that I added to the AzureManagedMachinePool reconciler, I think we are close to as good as we can get. The initial reconcile of AMMP only starts when control plane transitions to ready. Once ready, the reconciler will try to reach goal state persistently with requeues if it reaches what is deemed a transient error, a 404 for the VMSS other similar errors that will be resolved as infra comes online. I don't think this is the most elegant solution, but it seems to be resilient.

**Comment by devigned** [comment]
success 7?, 22mins.

File: azure/regional_baseuri.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	}, nil
+}
+
+// BaseURI return a regional base URI, like `https://{region}.management.azure.com`.
+func (a *baseURIAdapter) BaseURI() string {
+	if a == nil || a.parsedURL == nil || a.Region == "" {
+		return a.aliasAuth.BaseURI()
+	}
+
+	sansScheme := path.Join(fmt.Sprintf("%s.%s", a.Region, a.parsedURL.Host), a.parsedURL.Path)
```
Comment (alexeldeib) [inline azure/regional_baseuri.go:58]: Do we actually want path.join() (os filepath separator) here, versus a hardcoded slash? Wouldn’t this now be platform dependent?

Comment (alexeldeib) [reply azure/regional_baseuri.go:58]: Nvm, I might be thinking filepath.join()

**Comment by alexeldeib** [comment]
Before the PR which reconciled all AMMP with cluster creation, we would create the cluster, wait for it to reconcile fully, then trigger creation of nodepools which takes ~5 min. With that approach I would expect to see around 25min runs based on my previous testing.

since we don’t need that extra 5min anymore (all pools should be created), I wonder if we have room to squeeze the times lower somewhere, or if that’s just as good as it gets.

**Comment by alexeldeib** [comment]
Nice work with the URL scheme/host/path parsing

/lgtm

File: azure/regional_baseuri.go (hunk 2 above)
Comment (devigned) [reply azure/regional_baseuri.go:58]: I think we are good: https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/path/path.go;l=174

**Comment by devigned** [comment]
> Before the PR which reconciled all AMMP with cluster creation, we would create the cluster, wait for it to reconcile fully, then trigger creation of nodepools which takes ~5 min. With that approach I would expect to see around 25min runs based on my previous testing.
> 
> since we don’t need that extra 5min anymore (all pools should be created), I wonder if we have room to squeeze the times lower somewhere, or if that’s just as good as it gets.

I like where this is going! I think we need to do some tinkering now that tests are more stable.

File: exp/controllers/azuremanagedmachinepool_reconciler.go [hunk 3]
```diff
@@ -74,12 +74,21 @@ func (a *AgentPoolVMSSNotFoundError) Is(target error) bool {
 }
 
 // newAzureManagedMachinePoolService populates all the services based on input scope.
-func newAzureManagedMachinePoolService(scope *scope.ManagedControlPlaneScope) *azureManagedMachinePoolService {
+func newAzureManagedMachinePoolService(scope *scope.ManagedControlPlaneScope) (*azureManagedMachinePoolService, error) {
+	var authorizer azure.Authorizer = scope
+	if scope.ControlPlane != nil {
+		regionalAuthorizer, err := azure.WithRegionalBaseURI(scope, scope.ControlPlane.Spec.Location)
```
Comment (CecileRobertMichon) [inline exp/controllers/azuremanagedmachinepool_reconciler.go:80]: should we use `scope.Location()` instead?

File: exp/controllers/azuremanagedmachinepool_reconciler.go [hunk 4]
```diff
@@ -74,12 +74,21 @@ func (a *AgentPoolVMSSNotFoundError) Is(target error) bool {
...
+		}
+		authorizer = regionalAuthorizer
+	}
+
 	return &azureManagedMachinePoolService{
 		scope:         scope,
 		agentPoolsSvc: agentpools.New(scope),
-		scaleSetsSvc:  scalesets.NewClient(scope),
-	}
+		scaleSetsSvc:  scalesets.NewClient(authorizer),
```
Comment (CecileRobertMichon) [inline exp/controllers/azuremanagedmachinepool_reconciler.go:90]: nice

**Review by CecileRobertMichon — COMMENTED** [review]
looks great, just one comment about using the Location() func

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Comment by CecileRobertMichon** [comment]
Ok that's fair @ cherry-pick bot :) 

@devigned do you think we should cherry-pick this to release-1.0?

**Comment by devigned** [comment]
> @devigned do you think we should cherry-pick this to release-1.0?

/shrug

With the next release coming very soon and managed clusters being experimental, I lean toward dealing with the cherry pick merge conflict.

**Pull request merged**